src/data/finsum.lean

```diff
@@ -0,0 +1,749 @@
+were not mapped to `1`, then this could be refactored into a multiplicative version.
+
+## Tags
+
+finsum, finsum_in, univ', finiteness
+-/
+
+open_locale classical
+
+universes u v w
+
+-- TODO move to data.finset.something?
+/-- `set.to_finset' s` is the finset corresponding to `s` if `s` is finite, and the empty set
+    otherwise -/
+noncomputable def set.to_finset' {α : Type u} (s : set α) : finset α :=
+if h : s.finite then h.to_finset else ∅
+
+theorem set.to_finset'_eq_to_finset {α : Type u} {s : set α} [fintype s] :
+  s.to_finset' = s.to_finset :=
+begin
+  have h : s.finite := ⟨infer_instance⟩,
+  unfold set.to_finset',
+  rw dif_pos h,
+  unfold set.finite.to_finset,
+  congr',
+end
+
+/-!
+
+# finset.univ' -- noncomputable univ (empty set if infinite)
+
+-/
+
+namespace finset
+
+/-- `univ' α` is the finset corresponding to all of α if α is finite, and the empty set otherwise.
+Note that it is noncomputable -/
+noncomputable def univ' (α : Type u) : finset α :=
+if h : nonempty (fintype α) then (classical.choice h).elems else ∅
+
+variable {α : Type u}
+
+/-- `univ' α` is `finset.univ` if `α` is a fintype. -/
+@[simp] lemma univ'_eq_univ [h : fintype α] : univ' α = univ :=
+by convert (dif_pos (nonempty.intro h))
+
+lemma univ'_eq_empty (h : ¬ nonempty (fintype α)) : univ' α = ∅ :=
+dif_neg h
+
+lemma univ'_eq_empty' (h : fintype α → false) : univ' α = ∅ :=
+univ'_eq_empty $ not_nonempty_fintype.2 $ ⟨h⟩
+
+lemma mem_univ' [h : fintype α] (x : α) : x ∈ univ' α :=
+(@univ'_eq_univ _ h).symm ▸ mem_univ _
+
+@[simp] lemma mem_univ'' [h : nonempty (fintype α)] (x : α) : x ∈ univ' α :=
+@mem_univ' α (nonempty.some h) x
+
+end finset
+
+/-! # noncomputable finite sums -/
+
+section finsum
+
+variables {α : Type u} {M : Type v} [add_comm_monoid M]
+
+/-- Sum of `f x` as `x` ranges over the elements of the support of `f`, if it's finite. Zero
+  otherwise. -/
+noncomputable def finsum (f : α → M) : M :=
+(function.support f).to_finset'.sum f
+
+/-- Sum of `f x` for `x ∈ s`, if `s ∩ f.support` is finite. Zero if it's infinite.  -/
+noncomputable def finsum_in (s : set α) (f : α → M) : M :=
+finsum (λ x, if x ∈ s then f x else 0)
+
+localized "notation `∑ᶠ` binders `, ` r:(scoped:67 f, finsum f) := r"
+  in big_operators
+
+localized "notation `∑ᶠ` binders ` in ` s `, ` r:(scoped:67 f, finsum_in s f) := r"
+  in big_operators
+
+open_locale big_operators
+
+lemma finsum_def (f : α → M) :
+  ∑ᶠ i : α, f i = if h : (function.support f).finite then finset.sum (h.to_finset) f else 0 :=
+begin
+  split_ifs,
+  { unfold finsum set.to_finset',
+    rw dif_pos h },
+  { unfold finsum,
+    convert finset.sum_empty,
+    exact dif_neg h }
+end
+
+lemma finsum_in_def (s : set α) (f : α → M) :
+  ∑ᶠ i in s, f i = finsum (λ x, if x ∈ s then f x else 0) := rfl
+
+-- function.support doesn't import data.finsupp and vice versa
+-- so the below lemma can't go in either place
```

Comment:
I think that we should `import data.support` in `data.finsupp.basic` and move these lemmas to `data.finsupp.basic`.
This lemma (with LHS and RHS swapped) can be named `finsupp.coe_support`.

Reply:
I see that this lemma was added at `fun_support_eq` so I have removed it